Support after_breakout hook

k0kubun opened the pull request:
I want to execute a hook after execution of `exit` or `quit` (not all `:after_session`). Since `exit` command throws `:breakout` in `exit-all`, `Pry::Command#call_with_hooks` doesn't call `instance_exec` for `after_hooks`.

Is there any way to do it without this patch?

kyrylo commented:
Have you tried using Ruby's `at_exit`?

k0kubun commented:
No. That's because my interest is not in `Kernel#exit` but in pry's `exit` command. It just throws `:breakout` and does not exit interpreter, isn't it?

k0kubun commented:
Ah, I'm using pry on console of `rails s`. So after using `exit` command, the interpreter continues to work.
I want to execute command just after exiting pry's REPL with `exit` command.

strcmp commented:
it is correct that `exit-all` command will throw a `:breakout` symbol to interrupt the repl loop and return control to the caller of the `binding.pry` call or Pry.start, etc, and it may not be only the command that does. any pry command can throw it to basically terminate a repl session.

k0kubun commented:
> it may not be only the command that does. any pry command can throw it to basically terminate a repl session.

I see. Would it be OK to add `:after_breakout` hook? My usecase is replacing https://github.com/deivid-rodriguez/pry-byebug/commit/de89e05b6ec7c292e294754e3b293fbaa51dcb22 with a hook like that.

strcmp commented:
why not use `:after_exit` instead? it seems like that's what a breakout will always do, and having both seems be confusing, and it also appears as though `after_exit` just doesn't work at the moment?

strcmp commented:
oh sorry, nevermind that comment, it never existed. yeah i don't know, after_exit/breakout seems like a fine name. yeah code looks good to me, you pass the return value from catch() too :+1:

rf- commented:
This looks fine, but I wonder if you might want to include the `raise_up` case too? Otherwise byebug would continue running if someone uses the `raise-up` command, right?

rf- commented:
Also it would be good to see a spec or two for this behavior.

k0kubun commented:
> Also it would be good to see a spec or two for this behavior.

Added a spec for after_breakout in 29699a1c853747bc14bb7f76fb6acf4672ebb696.

> This looks fine, but I wonder if you might want to include the raise_up case too? Otherwise byebug would continue running if someone uses the raise-up command, right?

That's true. Added a support for after_raise_up in https://github.com/pry/pry/commit/e330b01c30f5fd113a83f5a37dbe2b5954f75c99.

rf- commented:
Wait, now I'm confused. Why can't you just use `after_session` for your use case?

k0kubun commented:
That's because `after_session`'s scope is too big. Since pry-byebug has custome pry processor, `after_session` does not suite for stopping Byebug. Processing pry-byebug's `next`, `step` or `finish` are done outside of `Pry::REPL#start` and we shouldn't stop Byebug for these cases.

k0kubun commented:
Since I no longer need this, closing. Thank you for your reviews.
